README.md
# neokeys

Get characters from a terminal keyboard, similar to `getch` in [curses][0] library.

[![On pub.dev][pub_img]][pub_url]
[![Code coverage][cov_img]][cov_url]
[![Github action status][gha_img]][gha_url]
[![Dartdocs][doc_img]][doc_url]
[![Style guide][sty_img]][sty_url]

[pub_url]: https://pub.dartlang.org/packages/neokeys
[pub_img]: https://img.shields.io/pub/v/neokeys.svg
[gha_url]: https://github.com/neo-dart/neokeys/actions
[gha_img]: https://github.com/neo-dart/neokeys/workflows/Dart/badge.svg
[cov_url]: https://codecov.io/gh/neo-dart/neokeys
[cov_img]: https://codecov.io/gh/neo-dart/neokeys/branch/main/graph/badge.svg
[doc_url]: https://www.dartdocs.org/documentation/neokeys/latest
[doc_img]: https://img.shields.io/badge/Documentation-neokeys-blue.svg
[sty_url]: https://pub.dev/packages/neodart
[sty_img]: https://img.shields.io/badge/style-neodart-9cf.svg

## Purpose

Dart's terminal support is fairly basic, with a [`Stdin`][1] class and
little more, particularly for _input_ (see [`examples/without_neokeys.dart`][2])

As a result creating something representing a rendering loop quite difficult:

```dart
import 'dart:io';

void main() {
  stdin
    ..echoMode = false
    ..lineMode = false;
  _exampleOfWaitingForQ();
}

void _exampleOfWaitingForQ() async {
  const qKey = 0x71;

  final buffer = <List<int>>[];

  bool isHit(int keyCode) {
    return buffer.any((keys) => keys.length == 1 && keys.first == keyCode);
  }

  stdin.listen((keys) {
    if (completer.isCompleted) {
      return;
    }

    buffer.add(keys);
  });

  const frames = Duration(milliseconds: 1000 ~/ 30);
  await for (final _ in Stream<void>.periodic(frames)) {
    if (isHit(qKey)) {
      stdin
        ..echoMode = true
        ..lineMode = true;
      return;
    }
  }
}
```

[0]: https://linux.die.net/man/3/getch
[1]: https://api.dart.dev/stable/2.17.1/dart-io/Stdin-class.html
[2]: examples/without_neokeys.dart

## Usage

The same code as above in [_purpose_](#purpose), using `neokeys`:

```dart
import 'dart:io';

import 'package:neokeys/neokeys.dart';

void main() {
  stdin
    ..echoMode = false
    ..lineMode = false;
  _exampleOfWaitingForQ();
}

void _exampleOfWaitingForQ() async {
  // Creates a smart listener with buffering.
  final input = stdin.neokeys();

  const frames = Duration(milliseconds: 1000 ~/ 30);
  await for (final _ in Stream<void>.periodic(frames)) {
    // Fully typed API with convenience methods.
    if (input.isPressed(Key.q)) {
      stdin
        ..echoMode = true
        ..lineMode = true;
      return;
    }
  }
}
```

## Contributing

**This package welcomes [new issues][issues] and [pull requests][fork].**

[issues]: https://github.com/neo-dart/neokeys/issues/new
[fork]: https://github.com/neo-dart/neokeys/fork

Changes or requests that do not match the following criteria will be rejected:

1. Common decency as described by the [Contributor Covenant][code-of-conduct].
2. Making this library brittle.
3. Adding platform-specific functionality.
4. A somewhat arbitrary bar of "complexity", everything should be _easy to use_.

[code-of-conduct]: https://www.contributor-covenant.org/version/1/4/code-of-conduct/


CHANGELOG.md
# CHANGELOG

## 0.1.1

- `<BufferedReader.isPressed>` support up to 6 codes to capture some patterns.

## 0.1.0

- Initial release, including a `BufferedKeys` class.


test/buffered_keys_test.dart
import 'dart:async';

import 'package:neokeys/neokeys.dart';
import 'package:test/test.dart';

void main() {
  const q = 0x71;
  const z = 0x7A;

  group('BufferedKeys.sync', () {
    late BufferedKeys keys;
    late List<List<int>> buffer;

    setUp(() {
      keys = BufferedKeys.sync(buffer = []);
    });

    test('should not indicate any key is pressed', () {
      expect(keys.isAnyPressed, isFalse);
    });

    test('should indicate any key is pressed', () {
      buffer.add([q]);

      expect(keys.isAnyPressed, isTrue);
    });

    test('should only indicate a specific key is pressed', () {
      buffer.add([q]);

      expect(keys.isPressed(q), isTrue);
      expect(keys.isPressed(z), isFalse);
    });

    test('should clear the buffer on demand', () {
      buffer.add([q]);
      keys.clear();

      expect(keys.isPressed(q), isFalse);
    });

    test('should clear the buffer when cancelled', () {
      buffer.add([q]);
      keys.cancel();

      expect(keys.isPressed(q), isFalse);
    });

    test('should check for a 2-code character', () {
      buffer.add([1, 2]);

      expect(keys.isPressed(1, 2), isTrue);
    });

    test('should check for a 3-code character', () {
      buffer.add([1, 2, 3]);

      expect(keys.isPressed(1, 2, 3), isTrue);
    });

    test('should check for a 4-code character', () {
      buffer.add([1, 2, 3, 4]);

      expect(keys.isPressed(1, 2, 3, 4), isTrue);
    });

    test('should check for a 5-code character', () {
      buffer.add([1, 2, 3, 4, 5]);

      expect(keys.isPressed(1, 2, 3, 4, 5), isTrue);
    });

    test('should check for a 6-code character', () {
      buffer.add([1, 2, 3, 4, 5, 6]);

      expect(keys.isPressed(1, 2, 3, 4, 5, 6), isTrue);
    });
  });

  group('BufferedKeys.async', () {
    late BufferedKeys keys;
    late StreamController<List<int>> buffer;

    Future<void> pumpEventQueue() => Future(() {});

    setUp(() {
      keys = BufferedKeys.async((buffer = StreamController()).stream);
    });

    test('should not indicate any key is pressed', () {
      expect(keys.isAnyPressed, isFalse);
    });

    test('should indicate any key is pressed', () async {
      buffer.add([q]);
      await pumpEventQueue();

      expect(keys.isAnyPressed, isTrue);
    });

    test('should only indicate a specific key is pressed', () async {
      buffer.add([q]);
      await pumpEventQueue();

      expect(keys.isPressed(q), isTrue);
      expect(keys.isPressed(z), isFalse);
    });

    test('should clear the buffer on demand', () async {
      buffer.add([q]);
      await pumpEventQueue();
      keys.clear();

      expect(keys.isPressed(q), isFalse);
    });

    test('should clear the buffer when cancelled', () async {
      buffer.add([q]);
      await pumpEventQueue();
      keys.cancel();

      expect(keys.isPressed(q), isFalse);
    });

    test('should check for a 2-code character', () async {
      buffer.add([1, 2]);
      await pumpEventQueue();

      expect(keys.isPressed(1, 2), isTrue);
    });

    test('should check for a 3-code character', () async {
      buffer.add([1, 2, 3]);
      await pumpEventQueue();

      expect(keys.isPressed(1, 2, 3), isTrue);
    });
  });
}


lib/neokeys.dart
/// Get characters from a terminal keyboard, similar to `getch` in [curses][].
///
/// Use [BufferedKeys] to buffer and allow synchronous recall of pressed keys.
///
/// [curses]: https://linux.die.net/man/3/getch
library neokeys;

import 'dart:async';

import 'package:meta/meta.dart';

/// A minimal buffer interface for _synchronously_ reading the state of keys.
///
/// When reading from the terminal (i.e. with `dart:io`), you have two choices:
/// 1. Block the program until a key is pressed (`stdin.readByteSync()`).
/// 2. Capture event codes as they are received (using `stdin` as a [Stream]).
///
/// [BufferedKeys] provides a synchronous API for the _state_ of key codes:
/// ```
/// void processInput(BufferedKeys keys) {
///   if (keys.isPressed(/* q */ 0x71)) {
///     print('"q" was pressed');
///   }
/// }
/// ```
///
/// The simplest implementation uses `dart:io` and [BufferedKeys.async], i.e.:
/// ```dart
/// import 'dart:io';
///
/// void main() async {
///   const frames = Duration(milliseconds: 1000 ~/ 30);
///   final keys = BufferedKeys.from(stdio);
///
///   stdout.writeln('Press any key to quit');
///
///   await for (final _ in Stream.periodic(frames)) {
///     if (keys.isAnyPressed) {
///       return;
///     }
///   }
/// }
/// ```
///
/// ## Lifecycle
///
/// This class on most platforms is **stateful**, and care should be taken to
/// avoid memory leaks or hangs in applications. In general, only two methods
/// ([clear] and [cancel]) must be used:
///
/// 1. After processing keys (i.e. after using [isPressed]), call [clear].
/// 2. When reading of keys is no longer necessary, call [cancel].
///
/// ```dart
/// // An example of a game loop lifecycle.
/// class GameLoop {
///   BufferedKeys keys;
///
///   /* ... */
///
///   void terminateLoop() {
///     keys.cancel();
///   }
///
///   void processInput() {
///     if (keys.isAnyPressed) {
///       stdout.writeln('Key pressed!');
///     }
///     keys.clear();
///   }
///
///   /* ... */
/// }
/// ```
abstract class BufferedKeys {
  /// @nodoc
  const BufferedKeys();

  /// Creates a buffer of key codes from a stream of key codes being received.
  ///
  /// This is a reasonable default from listening to `stdin`:
  /// ```dart
  /// import 'dart:io';
  ///
  /// void main() {
  ///   final keys = BufferedKeys.async(stdin);
  ///
  ///   /* ... */
  /// }
  /// ```
  factory BufferedKeys.async(Stream<List<int>> input) = _AsyncBufferedKeys;

  /// Creates a buffer of key codes using a sequence of codes as [input].
  ///
  /// This class is provided mostly for testing or stubbing purposes:
  /// ```dart
  /// void main() {
  ///   // Can be added to (buffer.add) to add codes to the buffer.
  ///   final buffer = <List<int>>[];
  ///   final keys = BufferedKeys.sync(buffer);
  ///
  ///   /* ... */
  /// }
  /// ```
  factory BufferedKeys.sync(List<List<int>> input) = _SyncBufferedKeys;

  /// Cancels the underlying event listeners, if any, for reading keys.
  ///
  /// Implementations that do not require cleanup should still call [clear].
  ///
  /// If using the terminal in raw mode, disable raw mode _before_ cancelling.
  @mustCallSuper
  void cancel() {
    clear();
  }

  /// Clears the buffer.
  ///
  /// This method should be invoked after checking keys using [isPressed] to
  /// clear the buffer between frames, otherwise keys will report being pressed
  /// that have been released and memory will grow indefinitely.
  void clear();

  /// Returns whether the provided key code exists in the buffer.
  ///
  /// May provide up to 6 codes in order to capture control characters:
  /// ```dart
  /// void processInput(BufferedKeys keys) {
  ///   final upArrowPressed = keys.isPressed(0x1B, 0x5B, 0x41);
  ///   // ...
  /// }
  /// ```
  bool isPressed(
    int code1, [
    int code2,
    int code3,
    int code4,
    int code5,
    int code6,
  ]);

  /// Returns whether any key is present in the buffer.
  ///
  /// Exactly what keys are considered "any" is platform specific.
  bool get isAnyPressed;
}

class _SyncBufferedKeys extends BufferedKeys {
  /// Buffered keycodes.
  final List<List<int>> _buffer;

  _SyncBufferedKeys(this._buffer);

  @override
  void clear() {
    _buffer.clear();
  }

  static const _notSet$ImplementationDetailDoNotUse = -1;

  @override
  bool isPressed(
    int code1, [
    int code2 = _notSet$ImplementationDetailDoNotUse,
    int code3 = _notSet$ImplementationDetailDoNotUse,
    int code4 = _notSet$ImplementationDetailDoNotUse,
    int code5 = _notSet$ImplementationDetailDoNotUse,
    int code6 = _notSet$ImplementationDetailDoNotUse,
  ]) {
    for (final keys in _buffer) {
      /// Returns whether the key at the provided index matches a code.
      ///
      /// - If code is undefined, a key must not be set for the index.
      /// - If code is specified, a key must be set and must be equal.
      bool hasMatch(int index, int code) {
        // An undefined code means we expect the buffer to be empty.
        if (code == _notSet$ImplementationDetailDoNotUse) {
          return index >= keys.length;
        }
        // A specified code means we expect the buffer to have the exact value.
        return index < keys.length && code == keys[index];
      }

      if (!hasMatch(0, code1) ||
          !hasMatch(1, code2) ||
          !hasMatch(2, code3) ||
          !hasMatch(3, code4) ||
          !hasMatch(4, code5) ||
          !hasMatch(5, code6)) {
        continue;
      }

      return true;
    }
    return false;
  }

  @override
  bool get isAnyPressed => _buffer.isNotEmpty;
}

class _AsyncBufferedKeys extends _SyncBufferedKeys {
  /// Subscription to the underlying stream.
  final StreamSubscription<void> _subscription;

  factory _AsyncBufferedKeys(Stream<List<int>> keys) {
    final buffer = <List<int>>[];
    return _AsyncBufferedKeys._(buffer, keys.listen(buffer.add));
  }

  _AsyncBufferedKeys._(
    List<List<int>> buffer,
    this._subscription,
  ) : super(buffer);

  @override
  void cancel() {
    _subscription.cancel();
    super.cancel();
  }
}


example/without_neokeys.dart
import 'dart:async';
import 'dart:io';

/// Run as `dart example/without_neokeys.dart N`.
///
/// 1. Reading from the keyboard synchronously
/// 2. Reading from the keyboard asynchronously
/// 3. Reading from the keyboard asynchronously with an "is hit"-like API
///
/// It *mostly* works, with caveats.
void main(List<String> args) async {
  var wasError = false;
  stdin
    ..echoMode = false
    ..lineMode = false;
  final program = args.length == 1 ? args.first : null;
  switch (program) {
    case '1':
      _readKeyboardSync();
      break;
    case '2':
      await _readKeyboardAsync();
      break;
    case '3':
      await _readKeyboardAsyncWithKeyHitLoop();
      break;
    default:
      stderr.writeln('Include a program parameter, i.e. 1 2 or 3');
      stderr.writeln('1 = readKeyboardSync');
      stderr.writeln('2 = readKeyboardAsync');
      stderr.writeln('3 = readKeyboardAsyncWithKeyHitLoop');
      wasError = true;
      break;
  }
  stdin
    ..echoMode = true
    ..lineMode = true;
  if (wasError) {
    exit(1);
  }
}

// Key code for `Q`.
const _qKey = 0x71;

String _debugKeyCode(int keyCode) {
  return '0x${keyCode.toRadixString(16).toUpperCase()} ($keyCode)';
}

/// It is super easy to read a single character code from [stdin].
///
/// What's hard:
///
/// - What if I want to read non-standard codes such as control keys?
/// - What if I *don't* want to synchronously block the entire program?
void _readKeyboardSync() {
  stdout.writeln('Press any (well, mostly any) key to quit:');

  final c = stdin.readByteSync();
  stdout.writeln('readByteSync(): ${_debugKeyCode(c)}');
}

/// It is super easy to read multiple character codes as a stream from [stdin].
///
/// What's hard:
///
/// - What if I build a rendering loop and want to process at certain time?
Future<void> _readKeyboardAsync() async {
  stdout.writeln('Press any key or key combination, or Q to quit');

  await for (final keys in stdin) {
    if (keys.length == 1 && keys.first == _qKey) {
      return;
    }

    final codes = keys.map(_debugKeyCode).toList();
    stdout.writeln('await for (... stdin): $codes');
  }
}

/// ... it's much more complicated to buffer keyboard reads until checked.
///
/// So, this is more or less the reason to use neokeys!
Future<void> _readKeyboardAsyncWithKeyHitLoop() async {
  late final StreamSubscription<void> read;
  final completer = Completer<void>();
  final buffer = <List<int>>[];

  void quit() {
    if (completer.isCompleted) {
      return;
    }
    read.cancel();
    completer.complete();
  }

  // This works... but what if you wanted to store timestamps too in order to
  // "detect" more accurately long a key was pressed? Do you just keep adding
  // and adding code?
  read = stdin.listen((keys) {
    if (completer.isCompleted) {
      return;
    }

    if (keys.length == 1 && keys.first == _qKey) {
      quit();
    }

    buffer.add(keys);
  });

  stdout.writeln('Press WASD to move, Q to quit');

  const aKey = 0x61;
  const dKey = 0x64;
  const sKey = 0x73;
  const wKey = 0x77;

  bool isHit(int keyCode) {
    return buffer.any((keys) => keys.length == 1 && keys.first == keyCode);
  }

  var xDirection = 0;
  var yDirection = 0;

  // This is not terribly efficient, but that's not the point.
  void processInput() {
    if (isHit(aKey)) {
      xDirection--;
    }
    if (isHit(dKey)) {
      xDirection++;
    }
    if (isHit(sKey)) {
      yDirection++;
    }
    if (isHit(wKey)) {
      yDirection--;
    }
    if (!completer.isCompleted && isHit(_qKey)) {
      quit();
    }
    if (xDirection > 1) {
      xDirection = 1;
    } else if (xDirection < -1) {
      xDirection = -1;
    }
    if (yDirection > 1) {
      yDirection = 1;
    } else if (yDirection < -1) {
      yDirection = -1;
    }
    buffer.clear();
  }

  var x = 50;
  var y = 50;

  void updateState() {
    x += xDirection;
    y += yDirection;

    if (x < 0) {
      x = 0;
    } else if (x > 100) {
      x = 100;
    }

    if (y < 0) {
      y = 0;
    } else if (y > 100) {
      y = 100;
    }
  }

  const frames = Duration(milliseconds: 1000 ~/ 30);

  // ignore: no_leading_underscores_for_local_identifiers
  await for (final _ in Stream<void>.periodic(frames)) {
    xDirection = yDirection = 0;

    processInput();
    updateState();

    if (xDirection != 0 || yDirection != 0) {
      stdout.writeln('You are standing at ($x, $y).');
    }

    if (completer.isCompleted) {
      return;
    }
  }

  await completer.future;
}


example/using_neokeys.dart
import 'dart:async';
import 'dart:io';

import 'package:neokeys/neokeys.dart';

/// Run as `dart example/using_neokeys.dart`.
void main() async {
  stdin
    ..echoMode = false
    ..lineMode = false;

  final keys = BufferedKeys.async(stdin);
  stdout.writeln('Press WASD to move, Q to quit');

  const aKey = 0x61;
  const dKey = 0x64;
  const sKey = 0x73;
  const wKey = 0x77;
  const qKey = 0x71;

  var xDirection = 0;
  var yDirection = 0;
  var shouldExit = false;

  void processInput() {
    if (keys.isPressed(aKey)) {
      xDirection--;
    }
    if (keys.isPressed(dKey)) {
      xDirection++;
    }
    if (keys.isPressed(sKey)) {
      yDirection++;
    }
    if (keys.isPressed(wKey)) {
      yDirection--;
    }
    if (keys.isPressed(qKey)) {
      shouldExit = true;
    }

    keys.clear();
  }

  var x = 50;
  var y = 50;

  void updateState() {
    x += xDirection;
    y += yDirection;

    if (x < 0) {
      x = 0;
    } else if (x > 100) {
      x = 100;
    }

    if (y < 0) {
      y = 0;
    } else if (y > 100) {
      y = 100;
    }
  }

  const frames = Duration(milliseconds: 1000 ~/ 30);

  // ignore: no_leading_underscores_for_local_identifiers
  await for (final _ in Stream<void>.periodic(frames)) {
    xDirection = yDirection = 0;

    processInput();
    updateState();

    if (xDirection != 0 || yDirection != 0) {
      stdout.writeln('You are standing at ($x, $y).');
    }

    if (shouldExit) {
      stdin
        ..echoMode = true
        ..lineMode = true;
      keys.cancel();
      break;
    }
  }
}
